File: src/assistant/wellbeing_coach.py
"""Mind wellbeing coaching responses and strategies."""
from typing import List, Dict, Optional
import random

from src.utils.config import config
from src.utils.logger import logger
from src.assistant.emotion_detector import EmotionDetector

class WellbeingCoach:
    """Provides mind wellbeing focused responses."""
    
    def __init__(self):
        """Initialize wellbeing coach."""
        self.emotion_detector = EmotionDetector()
        self.conversation_history: List[Dict] = []
        self.responses = self._build_response_templates()
        logger.info("Wellbeing coach initialized")
    
    def _build_response_templates(self) -> Dict[str, List[str]]:
        """Build response templates for different situations."""
        return {
            "greeting": [
                "Hello! I'm here to support you. How are you feeling today?",
                "Hi there! I'm Sprout, and I'm here to help with your mind wellbeing. What's on your mind?",
                "Welcome! I'm glad you're here. How can I support you today?",
                "Hello! Let's take a moment together. How are you doing?"
            ],
            "happy": [
                "That's wonderful to hear! I'm so glad you're feeling positive.",
                "It's beautiful to see you in a good space. What's contributing to this feeling?",
                "I'm happy for you! Celebrating these moments is important.",
                "That's fantastic! Positive moments like this are precious."
            ],
            "sad": [
                "I hear you, and I'm here with you. It's okay to feel this way.",
                "Thank you for sharing that with me. Your feelings are valid and important.",
                "I understand this is difficult. You're not alone in this.",
                "It takes courage to express these feelings. I'm here to support you."
            ],
            "anxious": [
                "I can sense you're feeling anxious. Let's take some deep breaths together.",
                "Anxiety can be overwhelming. Remember, this feeling will pass.",
                "You're safe here. Let's work through this together, one step at a time.",
                "I understand anxiety can be challenging. What helps you feel more grounded?"
            ],
            "tired": [
                "It sounds like you're feeling drained. Rest is important for your wellbeing.",
                "Being tired can affect how we feel. Have you been able to rest?",
                "I hear you're exhausted. Self-care is essential right now.",
                "Fatigue can be overwhelming. What would help you recharge?"
            ],
            "calm": [
                "It's beautiful that you're feeling calm and centered.",
                "I'm glad you're in a peaceful space. This is a good foundation.",
                "Calmness is a gift. How are you maintaining this sense of peace?",
                "That sense of calm is valuable. What's helping you stay grounded?"
            ],
            "encouragement": [
                "You're doing the best you can, and that's enough.",
                "Every step forward, no matter how small, is progress.",
                "You have strength within you, even when it doesn't feel that way.",
                "Remember, it's okay to not be okay. You're still moving forward.",
                "Your journey is unique, and you're navigating it with courage."
            ],
            "support": [
                "I'm here to listen and support you through this.",
                "You don't have to face this alone. I'm with you.",
                "Your wellbeing matters, and I'm here to help you nurture it.",
                "Together, we can explore ways to support your mind wellbeing."
            ],
            "closing": [
                "Take care of yourself. I'm here whenever you need support.",
                "Remember to be gentle with yourself. You're doing great.",
                "I'm glad we could connect. Take things one step at a time.",
                "You've got this. I'm here if you need to talk again."
            ]
        }
    
    def generate_response(
        self,
        user_input: str,
        conversation_context: Optional[List[Dict]] = None
    ) -> str:
        """
        Generate a wellbeing-focused response.
        
        Args:
            user_input: User's input text
            conversation_context: Previous conversation turns
            
        Returns:
            Response text
        """
        # Detect emotion
        emotion, confidence = self.emotion_detector.detect_emotion(user_input)
        wellbeing_level = self.emotion_detector.get_wellbeing_level(user_input)
        
        # Check for greetings
        if any(word in user_input.lower() for word in ["hello", "hi", "hey", "greetings"]):
            return random.choice(self.responses["greeting"])
        
        # Generate emotion-specific response
        if emotion in self.responses:
            base_response = random.choice(self.responses[emotion])
        else:
            base_response = random.choice(self.responses["support"])
        
        # Add encouragement if needed
        if wellbeing_level == "low" or emotion in ["sad", "anxious", "tired"]:
            encouragement = random.choice(self.responses["encouragement"])
            return f"{base_response} {encouragement}"
        
        return base_response
    
    def get_suggestions(self, emotion: str, wellbeing_level: str) -> List[str]:
        """
        Get wellbeing suggestions based on emotion and level.
        
        Args:
            emotion: Detected emotion
            wellbeing_level: Wellbeing level
            
        Returns:
            List of suggestions
        """
        suggestions = []
        
        if wellbeing_level == "low":
            suggestions = [
                "Take a few deep breaths - in through your nose, out through your mouth.",
                "Try a short walk, even just around your space.",
                "Remember to drink some water and take care of your basic needs.",
                "Consider writing down what you're feeling - it can help process emotions.",
                "You might want to try a brief meditation or mindfulness exercise."
            ]
        elif emotion == "anxious":
            suggestions = [
                "Try the 4-7-8 breathing technique: breathe in for 4, hold for 7, out for 8.",
                "Ground yourself by naming 5 things you can see, 4 you can touch, 3 you can hear.",
                "Remember that anxiety is temporary - this feeling will pass.",
                "Consider what's within your control right now, and focus on that."
            ]
        elif emotion == "tired":
            suggestions = [
                "Rest is important. Can you take a moment to pause?",
                "Consider what activities truly recharge you.",
                "Remember that rest is productive - you're not being lazy.",
                "Try to maintain a regular sleep schedule if possible."
            ]
        else:
            suggestions = [
                "Continue doing what's working for you.",
                "Consider what practices help maintain your sense of wellbeing.",
                "Remember to celebrate the positive moments.",
                "Stay connected with activities and people that support you."
            ]
        
        return suggestions
    
    def add_to_history(self, user_input: str, response: str, emotion: str):
        """Add conversation turn to history."""
        self.conversation_history.append({
            "user": user_input,
            "assistant": response,
            "emotion": emotion
        })
        
        # Keep history within limit
        if len(self.conversation_history) > config.max_conversation_history:
            self.conversation_history.pop(0)

File: Sources/Sprout/WellbeingCoach.swift
import Foundation

class WellbeingCoach: ObservableObject {
    @Published var currentMood: Mood = .neutral
    @Published var breathingExerciseActive = false
    @Published var sessionActive = false
    @Published var dailyStreak = 0
    @Published var weeklyProgress: [Date: Int] = [:]
    
    enum Mood: String, CaseIterable {
        case great = "Great"
        case good = "Good"
        case neutral = "Neutral"
        case low = "Low"
        case struggling = "Struggling"
        
        var emoji: String {
            switch self {
            case .great: return "🌟"
            case .good: return "😊"
            case .neutral: return "😐"
            case .low: return "😔"
            case .struggling: return "💙"
            }
        }
    }
    
    private let breathingExercises = [
        BreathingExercise(name: "4-7-8 Breathing", duration: 120, pattern: [4, 7, 8], emoji: "🌬️"),
        BreathingExercise(name: "Box Breathing", duration: 120, pattern: [4, 4, 4, 4], emoji: "📦"),
        BreathingExercise(name: "Deep Calm", duration: 180, pattern: [5, 5, 5], emoji: "🧘")
    ]
    
    struct BreathingExercise {
        let name: String
        let duration: Int // seconds
        let pattern: [Int] // inhale, hold, exhale, hold
        let emoji: String
    }
    
    init() {
        loadProgress()
        setupNotifications()
    }
    
    private func setupNotifications() {
        NotificationCenter.default.addObserver(
            forName: NSNotification.Name("UserMessageReceived"),
            object: nil,
            queue: .main
        ) { [weak self] notification in
            if let text = notification.object as? String {
                Task {
                    await self?.handleUserMessage(text)
                }
            }
        }
    }
    
    func handleUserMessage(_ text: String) async {
        let lowerText = text.lowercased()
        
        // Detect mood
        if lowerText.contains("great") || lowerText.contains("amazing") || lowerText.contains("wonderful") {
            currentMood = .great
        } else if lowerText.contains("good") || lowerText.contains("fine") || lowerText.contains("okay") {
            currentMood = .good
        } else if lowerText.contains("sad") || lowerText.contains("down") || lowerText.contains("low") {
            currentMood = .low
        } else if lowerText.contains("struggling") || lowerText.contains("difficult") || lowerText.contains("hard") {
            currentMood = .struggling
        }
        
        // Detect breathing exercise requests
        if lowerText.contains("breathing") || lowerText.contains("breathe") || lowerText.contains("calm") {
            await startBreathingExercise()
            return
        }
        
        // Detect meditation requests
        if lowerText.contains("meditate") || lowerText.contains("meditation") || lowerText.contains("mindful") {
            await startMeditationSession()
            return
        }
        
        // Generate supportive response
        let response = generateResponse(for: text)
        let emoji = getEmojiForResponse(response)
        
        await globalVoiceAssistant?.speak(response, emoji: emoji)
        
        // Show emoji
        NotificationCenter.default.post(
            name: NSNotification.Name("WellbeingEmoji"),
            object: emoji
        )
    }
    
    private func generateResponse(for text: String) -> String {
        let lowerText = text.lowercased()
        
        // Greetings
        if lowerText.contains("hello") || lowerText.contains("hi") || lowerText.contains("hey") {
            return "Hello! I'm here to support your mind wellbeing. How are you feeling today?"
        }
        
        // Mood responses
        if lowerText.contains("sad") || lowerText.contains("down") {
            return "I hear you. It's okay to feel this way. Would you like to try a breathing exercise together? It can help create a moment of calm."
        }
        
        if lowerText.contains("anxious") || lowerText.contains("worried") || lowerText.contains("stressed") {
            return "I understand that feeling. Let's take a moment to ground ourselves. Try taking three deep breaths with me."
        }
        
        if lowerText.contains("tired") || lowerText.contains("exhausted") {
            return "Rest is important for your wellbeing. Remember to be gentle with yourself. Would you like a short guided relaxation?"
        }
        
        if lowerText.contains("thank") {
            return "You're so welcome! I'm here whenever you need support. Remember, taking care of your mind is a journey, not a destination."
        }
        
        // Default supportive response
        return "I'm listening. Your feelings are valid. What would help you feel more grounded right now?"
    }
    
    private func getEmojiForResponse(_ response: String) -> String {
        if response.contains("breathing") || response.contains("breathe") {
            return "🌬️"
        }
        if response.contains("meditation") || response.contains("mindful") {
            return "🧘"
        }
        if response.contains("calm") || response.contains("relax") {
            return "🌊"
        }
        if response.contains("support") || response.contains("here") {
            return "💙"
        }
        return "🌱"
    }
    
    func startBreathingExercise() async {
        let exercise = breathingExercises.randomElement() ?? breathingExercises[0]
        breathingExerciseActive = true
        
        let message = "Let's do \(exercise.name). \(exercise.emoji) I'll guide you through it."
        await globalVoiceAssistant?.speak(message, emoji: exercise.emoji)
        
        // Guide through breathing pattern
        await guideBreathing(exercise)
    }
    
    private func guideBreathing(_ exercise: BreathingExercise) async {
        let pattern = exercise.pattern
        var cycle = 0
        
        while breathingExerciseActive && cycle < exercise.duration / 10 {
            // Inhale
            await globalVoiceAssistant?.speak("Breathe in... \(exercise.emoji)", emoji: exercise.emoji)
            try? await Task.sleep(nanoseconds: UInt64(pattern[0]) * 1_000_000_000)
            
            if pattern.count > 1 {
                // Hold
                await globalVoiceAssistant?.speak("Hold...", emoji: exercise.emoji)
                try? await Task.sleep(nanoseconds: UInt64(pattern[1]) * 1_000_000_000)
            }
            
            // Exhale
            await globalVoiceAssistant?.speak("Breathe out...", emoji: exercise.emoji)
            try? await Task.sleep(nanoseconds: UInt64(pattern.count > 2 ? pattern[2] : pattern[0]) * 1_000_000_000)
            
            if pattern.count > 3 {
                // Hold
                try? await Task.sleep(nanoseconds: UInt64(pattern[3]) * 1_000_000_000)
            }
            
            cycle += 1
        }
        
        breathingExerciseActive = false
        await globalVoiceAssistant?.speak("Great job! How do you feel now?", emoji: "🌱")
    }
    
    func startMeditationSession() async {
        sessionActive = true
        await globalVoiceAssistant?.speak("Let's take a moment for mindfulness. Find a comfortable position and close your eyes if you'd like.", emoji: "🧘")
        
        // Short guided meditation
        try? await Task.sleep(nanoseconds: 5_000_000_000)
        await globalVoiceAssistant?.speak("Notice your breath, without trying to change it. Just observe.", emoji: "🌊")
        
        try? await Task.sleep(nanoseconds: 10_000_000_000)
        await globalVoiceAssistant?.speak("If your mind wanders, that's okay. Gently bring your attention back to your breath.", emoji: "🌿")
        
        try? await Task.sleep(nanoseconds: 10_000_000_000)
        await globalVoiceAssistant?.speak("Take a moment to notice how you feel. You've done something wonderful for your mind wellbeing.", emoji: "✨")
        
        sessionActive = false
        updateProgress()
    }
    
    private func updateProgress() {
        let today = Calendar.current.startOfDay(for: Date())
        weeklyProgress[today, default: 0] += 1
        
        // Check for daily streak
        let yesterday = Calendar.current.date(byAdding: .day, value: -1, to: today)!
        if weeklyProgress[yesterday, default: 0] > 0 {
            dailyStreak += 1
        } else {
            dailyStreak = 1
        }
        
        saveProgress()
    }
    
    private func saveProgress() {
        // Save to UserDefaults or file
        UserDefaults.standard.set(dailyStreak, forKey: "dailyStreak")
    }
    
    private func loadProgress() {
        dailyStreak = UserDefaults.standard.integer(forKey: "dailyStreak")
    }
}

// Global reference - will be set by SproutApp
var globalVoiceAssistant: VoiceAssistant?

